### PR TITLE
[Python] Fall back to DOUBLE for unsupported DECIMAL widths

### PR DESCRIPTION
**Pull request by Tishj**
This PR is a follow up to #4690 

Previously we would fall back to VARCHAR if the width was too big to convert to DECIMAL, now we fallback to DOUBLE instead.

### CODE REVIEW

File: tools/pythonpkg/src/python_objects.cpp [hunk 1]
```diff
@@ -118,10 +120,12 @@ void PyDecimal::SetExponent(py::handle &exponent) {
 	ExponentNotRecognized();
 }
 
-static void UnsupportedWidth(uint16_t width) {
-	throw ConversionException(
-	    "Failed to convert to a DECIMAL value with a width of %d because it exceeds the max supported with of %d",
-	    width, Decimal::MAX_WIDTH_INT128);
+static bool WidthIsCorrect(int32_t width) {
```
Comment (pdet) [inline tools/pythonpkg/src/python_objects.cpp:123]: This function name is weird, maybe `WidthFitsInDecimal`?

File: tools/pythonpkg/src/python_objects.cpp [hunk 2]
```diff
@@ -138,14 +142,27 @@ Value PyDecimalCastSwitch(PyDecimal &decimal, uint8_t width, uint8_t scale) {
 	return OP::template Operation<int16_t>(decimal.signed_value, decimal.digits, width, scale);
 }
 
+//! Wont fit in a DECIMAL, fall back to DOUBLE
+static Value FallbackToDouble(py::handle &obj) {
```
Comment (pdet) [inline tools/pythonpkg/src/python_objects.cpp:146]: CastToDouble?

File: tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py [hunk 3]
```diff
@@ -581,6 +581,28 @@ def test_numeric_decimal_varying_sizes(self):
         print(reference)
         print(conversion)
 
+    def test_numeric_decimal_fallback_to_double(self):
```
Comment (pdet) [inline tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py:584]: Can you also add tests with a  column with a mix of decimals widths? 
I.e., ones that fit and some that don't

**Review by pdet — COMMENTED** [review]
Sorry for my belated review, was a busy week.

I think it looks good, just some minor comments.

**Comment by Tishj** [comment]
I've implemented the feedback, I actually overlooked a scenario which required a WidthFitsInDecimal check, the new test luckily caught that :P

File: tools/pythonpkg/src/python_objects.cpp [hunk 4]
```diff
@@ -138,25 +142,41 @@ Value PyDecimalCastSwitch(PyDecimal &decimal, uint8_t width, uint8_t scale) {
 	return OP::template Operation<int16_t>(decimal.signed_value, decimal.digits, width, scale);
 }
 
+//! Wont fit in a DECIMAL, fall back to DOUBLE
```
Comment (pdet) [inline tools/pythonpkg/src/python_objects.cpp:145]: Just //

File: tools/pythonpkg/src/python_objects.cpp [hunk 5]
```diff
@@ -138,25 +142,41 @@ Value PyDecimalCastSwitch(PyDecimal &decimal, uint8_t width, uint8_t scale) {
...
+	}
 	switch (exponent_type) {
 	case PyDecimalExponentType::EXPONENT_SCALE: {
 		uint8_t scale = exponent_value;
-		if (width > Decimal::MAX_WIDTH_INT128) {
-			UnsupportedWidth(width);
-		}
+		D_ASSERT(WidthFitsInDecimal(width));
 		if (scale > width) {
 			//! Values like '0.001'
```
Comment (pdet) [inline tools/pythonpkg/src/python_objects.cpp:167]: //

File: tools/pythonpkg/src/python_objects.cpp [hunk 6]
```diff
@@ -138,25 +142,41 @@ Value PyDecimalCastSwitch(PyDecimal &decimal, uint8_t width, uint8_t scale) {
...
 }
 
+//! Wont fit in a DECIMAL, fall back to DOUBLE
+static Value CastToDouble(py::handle &obj) {
+	string converted = py::str(obj);
+	string_t decimal_string(converted);
+	double double_val;
+	bool try_cast = TryCast::Operation<string_t, double>(decimal_string, double_val, true);
+	if (!try_cast) {
+		DoubleConversionFail(converted);
```
Comment (pdet) [inline tools/pythonpkg/src/python_objects.cpp:152]: Add test to exception

File: tools/pythonpkg/src/python_objects.cpp [hunk 7]
```diff
@@ -138,25 +142,41 @@ Value PyDecimalCastSwitch(PyDecimal &decimal, uint8_t width, uint8_t scale) {
...
+		}
 		return PyDecimalCastSwitch<PyDecimalScaleConverter>(*this, width, scale);
 	}
 	case PyDecimalExponentType::EXPONENT_POWER: {
 		uint8_t scale = exponent_value;
 		width += scale;
-		if (width > Decimal::MAX_WIDTH_INT128) {
-			UnsupportedWidth(width);
+		if (!WidthFitsInDecimal(width)) {
+			return CastToDouble(obj);
```
Comment (pdet) [inline tools/pythonpkg/src/python_objects.cpp:179]: Add test that covers this

**Review by pdet — CHANGES_REQUESTED** [review]
Hey @Tishj , thanks for adding my requests already!

Some small nitpicks and asking for a couple more tests :-), then I think it's good to go.

**Comment by Tishj** [comment]
@pdet Could you check if this is ready to be merged? The Regression Test failing is a spurious HTTP failure

File: tools/pythonpkg/src/python_objects.cpp [hunk 8]
```diff
@@ -118,10 +120,12 @@ void PyDecimal::SetExponent(py::handle &exponent) {
...
-static void UnsupportedWidth(uint16_t width) {
-	throw ConversionException(
-	    "Failed to convert to a DECIMAL value with a width of %d because it exceeds the max supported with of %d",
-	    width, Decimal::MAX_WIDTH_INT128);
+static bool WidthFitsInDecimal(int32_t width) {
+	return width >= 0 && width <= Decimal::MAX_WIDTH_DECIMAL;
+}
+
+static void DoubleConversionFail(const string &decimal) {
+	throw ConversionException("Failed to fall back to a DOUBLE for DECIMAL '%s'", decimal);
```
Comment (pdet) [inline tools/pythonpkg/src/python_objects.cpp:128]: It seems that this error is never triggered by tests, can we have a test that triggers it?

File: tools/pythonpkg/src/python_objects.cpp [hunk 9]
```diff
@@ -118,10 +120,12 @@ void PyDecimal::SetExponent(py::handle &exponent) {
 	ExponentNotRecognized();
```
Comment (pdet) [inline tools/pythonpkg/src/python_objects.cpp:120]: It seems no test triggers this function call, can we have a test that triggers it?

**Review by pdet — CHANGES_REQUESTED** [review]
A couple of coverage tests missing 

https://app.codecov.io/gh/duckdb/duckdb/pull/4749

File: tools/pythonpkg/src/python_objects.cpp (hunk 8 above)
Comment (Tishj) [reply tools/pythonpkg/src/python_objects.cpp:128]: Ah yes, weird that the compiler doesn't complain about this unused static function.
I looked into the code this is calling, and experimented with it a bunch to make it fail, but I couldn't get it to fail, so I have changed the error check into an assertion. I'll remove this function

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**